### PR TITLE
Fix `tesseract --list-langs` and replace several GenericVector

### PR DESCRIPTION
**Pull request by stweil**

### CODE REVIEW

File: src/api/baseapi.cpp [hunk 1]
```diff
@@ -1903,7 +1909,9 @@ void TessBaseAPI::End() {
   delete block_list_;
   block_list_ = nullptr;
   if (paragraph_models_ != nullptr) {
-    paragraph_models_->delete_data_pointers();
+    for (auto model : *paragraph_models_) {
+      delete model;
```
Comment (egorpugin) [inline src/api/baseapi.cpp:1913]: Not good.
We should consider unique ptrs or something like this here.
I also tried to replace `delete_data_pointers()`, but noticed that it frees memory of elements and reverted changes locally.

There are several common patterns over the codebase. I think those places must be upgraded uniformely.

**Comment by egorpugin** [comment]
At the end I think I'm against raw `delete`.
Vector changes are fine except for eliminating `delete_data_pointers()`

File: src/api/baseapi.cpp [hunk 2]
```diff
@@ -278,6 +278,9 @@ void TessBaseAPI::SetInputName(const char* name) {
 
 /** Set the name of the output files. Needed only for debugging. */
 void TessBaseAPI::SetOutputName(const char* name) {
+  if (name == nullptr) {
+    name = "";
```
Comment (egorpugin) [inline src/api/baseapi.cpp:282]: This will be solved if we allow std::string everywhere.

File: include/tesseract/baseapi.h [hunk 3]
```diff
@@ -815,7 +815,7 @@ class TESS_API TessBaseAPI {
   EquationDetect* equ_detect_;     ///< The equation detector.
   FileReader reader_;              ///< Reads files from any filesystem.
   ImageThresholder* thresholder_;  ///< Image thresholding module.
-  GenericVector<ParagraphModel*>* paragraph_models_;
+  std::list<ParagraphModel*>* paragraph_models_;
```
Comment (egorpugin) [inline include/tesseract/baseapi.h:818]: Also suspicious.
If vector worked here, why change to list?
Vector merges are find using .insert(end(), v2.begin(), v2.end());

Comment (stweil) [reply include/tesseract/baseapi.h:818]: The code removes elements. That's expensive for `std::vector`, but cheap for `std::list`.

File: src/api/baseapi.cpp (hunk 1 above)
Comment (stweil) [reply src/api/baseapi.cpp:1913]: My first priority is currently getting a fully working git master again. A lot of code can be simplified or rewritten, but that can be done later.

File: include/tesseract/baseapi.h (hunk 3 above)
Comment (egorpugin) [reply include/tesseract/baseapi.h:818]: I see. Legit then.
But I propose to optimize things at the end.
We try to update 1-to-1 code to STL containers first. Optimization will be at the very end.

**Comment by stweil** [comment]
https://github.com/stweil/tesseract/tree/new-api is based on this pull request and completely removes `genericvector.h` and `strngs.h` from the public Tesseract API.

**Comment by egorpugin** [comment]
> completely removes genericvector.h and strngs.h

I see.
In this case I think you could merge it and we'll fix issues here in master together?
I do not want to deny or keep such changes in review for a long time.

**Comment by egorpugin** [comment]
We have unittests, we have base line for my changes (that alpha tag you made), so we can always compare things and fix new errors.

**Comment by stweil** [comment]
This pull request is part of the more complete pull request #3198, so I close it.

**Pull request closed without merging**